### ISSUE TITLE
Question about loratorch.register_model_param_after_backward(model)

### ISSUE THREAD

**Issue by wjm-wjm** [Baijiong-Lin/LoRA-Torch#10]
Hi, I have a question about loratorch.register_model_param_after_backward(model). Why need run loratorch.register_model_param_after_backward(model) after every backward process，is it ok to run this line of code when need to save checkpoint? Since it just change the weight type from torch.tensor to nn.Parameter ?

**Comment by Baijiong-Lin** [comment]
please refer to #7.

**Comment by wjm-wjm** [comment]
Hi, I find an another situation. If I set model.eval() and then save model, the weight of saving model would be added lora parameters. After that, if I try to load model and set model.eval() to inference, the lora parameters would be added twice. However, loralib not. (named_model1 is the saving model when I set model.eval(), named_model2 is the original model)

<img width="486" alt="Image" src="https://github.com/user-attachments/assets/c7833f47-b8e6-4249-8e77-745bfa547781" />

<img width="596" alt="Image" src="https://github.com/user-attachments/assets/8c733d9a-3429-439f-8be3-1fbfbae2e331" />

**Comment by Baijiong-Lin** [comment]
when calling model.eval(), lora will be merged into the model weights. loralib is also similar (https://github.com/microsoft/LoRA/blob/main/loralib/layers.py#L141).

**Comment by wjm-wjm** [comment]
However, I find that loralib will save "self.merged=True", but loratorch not

> when calling model.eval(), lora will be merged into the model weights. loralib is also similar (https://github.com/microsoft/LoRA/blob/main/loralib/layers.py#L141).

**Comment by Baijiong-Lin** [comment]
https://github.com/Baijiong-Lin/LoRA-Torch/blob/main/loratorch/layers.py#L110

**Comment by wjm-wjm** [comment]
Thanks, I may find mistake. That is if I save model at a eval mode, I must load checkpoint at a eval mode, which make "self.merged=True" consistent

**Comment by Baijiong-Lin** [comment]
whether saving the model after calling model.eval() depends on your experimental setting (if you execute model.eval() before saving the model, the saved weights are merged with lora; otherwise not). usually, we do not execute it before saving.

**Comment by wjm-wjm** [comment]
> whether saving the model after calling model.eval() depends on your experimental setting (if you execute model.eval() before saving the model, the saved weights are merged with lora; otherwise not). usually, we do not execute it before saving.

I set model.eval() and then save checkpoint because I need to do validation of the model for every training epoch. All in all, much thanks for your kind reply !

**Comment by Baijiong-Lin** [comment]
I think the following process is more popular?

```
# save model after training
torch.save(model)
torch.load(model)
model.eval() # lora will be merged here
model.forward() # start evaluation

# continually train lora after evaluation
model.train()
model.forward() # training
```

executing model.train() after evaluation will return the lora weights from model weights, and then you can continually train the lora.

**Comment by wjm-wjm** [comment]
> I think the following process is more popular?
> 
> ```
> # save model after training
> torch.save(model)
> torch.load(model)
> model.eval() # lora will be merged here
> model.forward() # start evaluation
> 
> # continually train lora after evaluation
> model.train()
> model.forward() # training
> ```
> 
> executing model.train() after evaluation will return the lora weights from model weights, and then you can continually train the lora.

Specifically, I need to compare the evaluation result to save model. Thus I may change the code into following:

```
model.eval() # lora will be merged here
model.forward() # start evaluation
# save model according to evaluation result
model.train()
torch.save(model)
# continually train lora after evaluation
torch.load(model)
model.forward() # training
```

**Comment by Baijiong-Lin** [comment]
seems good
